napalm/junos/junos.py

```diff
@@ -233,11 +233,21 @@ def _detect_config_format(self, config):
         return fmt
 
     def _load_candidate(self, filename, config, overwrite):
+        fmt = None
         if filename is None:
             configuration = config
         else:
             with open(filename) as f:
-                configuration = f.read()
+                try:
+                    configuration = etree.parse(f).getroot()
```

Comment:
For this to work the `f` file handle needs to be kept around, and the way to load the XML is a bit different when loading from file or loading from string (`etree.parse(fh).getroot()` or `etree.XML(conf_string)`), so I don't see a clean way to handle all cases in a simple way.

It could be modified though to first read the file, detect format, and parse XML from file handle only when approriate. In that case I would do something like this:
```
    def _load_candidate(self, filename, config, overwrite):
        if filename is None:
            configuration = config
            fmt = self._detect_config_format(configuration)
            if fmt == "xml":
                # Note: configuration cannot have encoding information, since it is already in string-format.
                configuration = etree.XML(configuration)
        else:
            with open(filename) as f:
                configuration = f.read()
                fmt = self._detect_config_format(configuration)
                if fmt == "xml":
                    # XML header encoding is evaluated, if present.
                    configuration = etree.parse(f).getroot()

        if (
            not self.lock_disable
            and not self.session_config_lock
            and not self.config_private
        ):
            # if not locked during connection time, will try to lock
            self._lock()

        try:
            if self.config_private:
                try:
                    self.device.rpc.open_configuration(private=True, normalize=True)
                except RpcError as err:
                    if str(err) == "uncommitted changes will be discarded on exit":
                        pass

            self.device.cu.load(
                configuration,
                format=fmt,
                overwrite=overwrite,
                ignore_warning=self.ignore_warning,
            )
        except ConfigLoadError as e:
            if self.config_replace:
                raise ReplaceConfigException(e.errs)
            else:
                raise MergeConfigException(e.errs)

```
What do you think?

Reply:
Yes, let's try this way.